SIMSBIOHUB-569/589: Add Method Techniques and Manage Sampling Information Page

mauberti-bc opened the pull request:
## Links to Jira Tickets

https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-589
https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-569

## Description of Changes

- New page to manage both sampling sites and techniques.
  - Techniques are created under a survey, and are applied to a sampling method (as part of creating/editing sampling sites).

## Testing Notes

- CRUD for sampling sites
- CRUD for techniques

mauberti-bc committed:
wip: migration to add techinques

mauberti-bc committed:
wip

mauberti-bc committed:
insert lookup values for techniques

mauberti-bc committed:
wip: technique service and repo

mauberti-bc committed:
view techniques for survey

LouisThedroux committed:
added values into existing structure, new enum type needs checking but works

LouisThedroux committed:
more technique attributes added

mauberti-bc committed:
method technique attribute select: broken

LouisThedroux committed:
adding more options, editing syntax and formatting

mauberti-bc committed:
wip: rework sampling site page and adding methods to sites

LouisThedroux committed:
finished adding core quant and qual attributes. Starting on adding options

LouisThedroux committed:
Merge branch 'method-technique' of https://github.com/bcgov/biohubbc into method-technique

LouisThedroux committed:
more attribute options

mauberti-bc committed:
formik errors on sampling site page for sample periods

mauberti-bc committed:
sampling method techniques on edit sample site page

mauberti-bc committed:
wip: replace method_technique_id with technique object in sample method response

mauberti-bc committed:
merge in dev

mauberti-bc committed:
fix types: include technique object in sample location get request but not updates nor inserts

mauberti-bc committed:
update observation table to include technique name

mauberti-bc committed:
wip: accordion card on manage sampling site page

mauberti-bc committed:
merge latest

LouisThedroux committed:
added method lookup attributes

mauberti-bc committed:
techniqueAttributeSelect is causing erroneous rerenders on value change

NickPhura committed:
Update sampling method skeletons

NickPhura committed:
Merge branch 'method-technique' of https://github.com/bcgov/biohubbc into method-technique

NickPhura committed:
Remove duplicate file

NickPhura committed:
Fix technique attributes form controls

NickPhura committed:
Run formatter

LouisThedroux committed:
added qualitative options

NickPhura committed:
Update technique attributes utils

mauberti-bc committed:
wip: edit technique page

mauberti-bc committed:
merge latest

LouisThedroux committed:
Added more attractants

LouisThedroux committed:
added comment for file explanation

LouisThedroux committed:
Merge remote-tracking branch 'origin/method-technique' into method-technique

mauberti-bc committed:
wip: resolve rerendering issues with technique attributes

mauberti-bc committed:
merge latest

mauberti-bc committed:
move units from technique_attribute_quantitative to method_tehcnique_attribute_quant

mauberti-bc committed:
wip: delete technique

LouisThedroux committed:
added min, max, unit to method_lookup_attribute_quantitative

LouisThedroux committed:
added more options and cleaned up some other formatting

mauberti-bc committed:
add, edit, and delete techniques quant and qual attributes and attractants

mauberti-bc committed:
Merge branch 'method-technique' of github.com:bcgov/biohubbc into method-technique

mauberti-bc committed:
wip: fix openapi spec for techniques

mauberti-bc committed:
db triggers for validating technique attributes & openapi schemas

mauberti-bc committed:
validate incoming attributes for techniques

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into method-technique

NickPhura committed:
Re-arrange components

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into method-technique

NickPhura committed:
Misc tweaks

NickPhura committed:
Misc updates

NickPhura committed:
Misc updates

NickPhura committed:
Misc updates

NickPhura committed:
Misc updates

NickPhura committed:
Merge branch 'method-technique' of https://github.com/bcgov/biohubbc into method-technique

NickPhura committed:
Updates

NickPhura committed:
Remove temp file

NickPhura committed:
More updates

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into method-technique

NickPhura committed:
fix merge issues

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into method-technique

NickPhura committed:
fix bulk technique menu

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into method-technique

NickPhura committed:
WIP

NickPhura committed:
Fix code after removing experimental react-query changes for sampling information.

NickPhura committed:
Fix linting, formatting, tests.

MacDeluca commented:
1. Deleting techniques fails "rows was null or undefined, expected rows != null"
2. Deleting multiple techniques has no effect
3. In add sampling site page, add sampling method dialog renders with horizontal scroll on half view port
4. (Maybe out of scope) the sampling sites cards are rendering strangely

![image](https://github.com/user-attachments/assets/4438c5f9-2f76-4f01-81e8-63fce2c9c3f1)

MacDeluca commented on api/src/paths/codes.ts:
```diff
@@ -349,6 +350,25 @@ GET.apiDoc = {
                   }
                 }
+              },
+              attractants: {
```
Title / description specs would be helpful here

NickPhura replied:
Added some stuff, though I probably need to go through and do the same for all of the codes, maybe as tech debt separately.

MacDeluca commented on api/src/paths/codes.ts:
```diff
@@ -349,6 +350,25 @@ GET.apiDoc = {
+                  additionalProperties: false,
+                  required: ['id', 'name', 'description'],
+                  properties: {
+                    id: {
```
minimum spec

NickPhura replied:
added

MacDeluca commented on api/src/paths/reference/get/technique-attribute.ts:
```diff
@@ -0,0 +1,173 @@
+                      },
+                      min: {
+                        type: 'integer',
+                        nullable: true
```
minimum spec

NickPhura replied:
negative values might need to be allowed

MacDeluca commented on api/src/paths/reference/get/technique-attribute.ts:
```diff
@@ -0,0 +1,173 @@
+                        type: 'integer',
+                        nullable: true
+                      },
+                      max: {
```
minimum spec

MacDeluca commented on api/src/paths/reference/get/technique-attribute.ts:
```diff
@@ -0,0 +1,173 @@
+              additionalProperties: false,
+              properties: {
+                method_lookup_id: {
+                  type: 'number'
```
minimum spec

NickPhura replied:
added

MacDeluca commented on api/src/repositories/attractants-repository.ts:
```diff
@@ -0,0 +1,165 @@
+  }
+
+  /**
+   * Delete all technique attractants.
```
'Delete all technique attractants of a survey'

NickPhura replied:
updated

MacDeluca commented on api/src/repositories/technique-attribute-repository.ts:
```diff
@@ -0,0 +1,529 @@
+  method_lookup_attribute_qualitative_id: string;
+}
+
+// export interface IQuantitativeAttributeRecord {
```
can we drop this dead code?💀

NickPhura replied:
Yes! I noticed there was some redundancy before, but must have forgotten to remove the old interfaces. Fixed

MacDeluca commented on api/src/repositories/technique-attribute-repository.ts:
```diff
@@ -0,0 +1,529 @@
+  max: z.number().nullable()
+});
+
+// export interface ITechniqueAttributesObject {
```
💀

NickPhura replied:
Removed

MacDeluca commented on api/src/repositories/technique-attribute-repository.ts:
```diff
@@ -0,0 +1,529 @@
+  options: z.array(TechniqueAttributeQualitativeOption)
+});
+
+// export interface IGetTechniqueAttributes {
```
💀

NickPhura replied:
removed

MacDeluca commented on api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/{surveySampleSiteId}/sample-method/index.ts:
```diff
@@ -94,7 +94,7 @@ GET.apiDoc = {
                     },
-                    method_lookup_id: {
+                    method_technique_id: {
                       type: 'integer'
```
min

NickPhura replied:
Updated, added some other missing fields to the spec as well.

MacDeluca commented on api/src/services/technique-service.ts:
```diff
@@ -0,0 +1,113 @@
+import { AttractantService } from './attractants-service';
+import { DBService } from './db-service';
+
+export class TechniqueService extends DBService {
```
jsdoc

NickPhura replied:
Updated

MacDeluca commented on app/src/assets/images/observations-overlay.png:
🐻

NickPhura replied:
🤷‍♂️

MacDeluca commented on app/src/components/dialog/ErrorDialog.tsx:
```diff
@@ -93,21 +96,25 @@ export const ErrorDialog = (props: IErrorDialogProps) => {
+              </Collapse>
+            </Box>
+          ) : (
+            <></>
```
Does this need to be a fragment for the stack? Or does null work?

NickPhura replied:
changed to `null`

MacDeluca commented on app/src/components/fields/AutocompleteField.tsx:
```diff
@@ -66,41 +73,80 @@ const AutocompleteField = <T extends string | number>(props: IAutocompleteField<
+            {...params}
+            required={props.required}
+            label={props.label}
+            InputLabelProps={
```
Remove prop?

NickPhura replied:
yep, removed

MacDeluca commented on app/src/components/fields/DateTimeFields.tsx:
```diff
@@ -75,8 +91,8 @@ export const DateTimeFields = <FormikPropsType,>(props: IDateTimeFieldsProps<For
-                error: get(touched, dateName) && Boolean(get(errors, dateName)),
-                helperText: get(touched, dateName) && get(errors, dateName),
+                error: get(touched, dateName) && (dateError || Boolean(get(errors, dateName))),
+                helperText: get(touched, dateName) && (dateHelperText || get(errors, dateName)),
```
Should we flip 'dateHelperText' and 'get(errors, dateName)' so the error will render if we are providing the dateHelperText?

```js
helperText: get(touched, dateName) && (get(errors, dateName) || dateHelperText)
```

NickPhura replied:
We actually never use dateHelperText anywhere, so I might just remove it

MacDeluca commented on app/src/components/fields/DateTimeFields.tsx:
```diff
@@ -120,8 +136,8 @@ export const DateTimeFields = <FormikPropsType,>(props: IDateTimeFieldsProps<For
-                error: get(touched, timeName) && Boolean(get(errors, timeName)),
-                helperText: get(touched, timeName) && get(errors, timeName),
+                error: get(touched, timeName) && (timeError || Boolean(get(errors, timeName))),
+                helperText: get(touched, timeName) && (timeHelperText || get(errors, timeName)),
```
same here

NickPhura replied:
We actually never use dateHelperText anywhere, so I might just remove it

MacDeluca commented on app/src/features/surveys/sampling-information/manage/SamplingSiteManageHeader.tsx:
```diff
@@ -0,0 +1,38 @@
+  survey_name: string;
+}
+
+export const SamplingSiteManageHeader = (props: SamplingSiteManageHeaderProps) => {
```
jsdoc

NickPhura replied:
added

MacDeluca commented on app/src/features/surveys/sampling-information/manage/SamplingSiteManagePage.tsx:
```diff
@@ -0,0 +1,32 @@
+import { SamplingTechniqueContainer } from 'features/surveys/sampling-information/techniques/SamplingTechniqueContainer';
+import { useProjectContext, useSurveyContext } from 'hooks/useContext';
+
+export const SamplingSiteManagePage = () => {
```
jsdoc

NickPhura replied:
Added

MacDeluca commented on app/src/features/surveys/sampling-information/methods/components/SamplingMethodForm.tsx:
```diff
@@ -0,0 +1,143 @@
+                subText: option.description ?? undefined
+              })) ?? []
+            }></SelectWithSubtextField>
+          {/* <AutocompleteField
```
💀

NickPhura replied:
Purged.

MacDeluca committed:
fix: updated some nit picks

NickPhura committed:
- Fix missing project permission on sampling endpoints
- Add technique bulk delete endpoint and wire up in frontend
- Update technique endpoints
- Add technique endpoint unit tests
- Remove commended code

NickPhura committed:
Merge branch 'method-technique' of https://github.com/bcgov/biohubbc into method-technique

NickPhura committed:
Merge branch 'dev' into method-technique

NickPhura committed:
Add backend unit tests for technique related code.

NickPhura committed:
Fix dialog width

NickPhura commented:
> 1. Deleting techniques fails "rows was null or undefined, expected rows != null"
> 
>     2. Deleting multiple techniques has no effect
> 
>     3. In add sampling site page, add sampling method dialog renders with horizontal scroll on half view port
> 
>     4. (Maybe out of scope) the sampling sites cards are rendering strangely
> 
> 

@MacQSL (1, 2, and 3) should be fixed. I didnt address 4. Will probably leave that for a separate UI polish ticket.

NickPhura committed:
Merge branch 'dev' into method-technique

MacDeluca reviewed (approved):
😶‍🌫️

NickPhura committed:
ignore-skip

NickPhura committed:
ignore-skip
